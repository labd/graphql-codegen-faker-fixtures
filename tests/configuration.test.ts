import { plugin } from "../dist";
import { buildSchema } from "graphql/utilities";
import { parse } from "graphql";

const schema = /* GraphQL */ `
  type Person {
    id: ID!
    name: String
    age: Int
    address: String
  }

  type Animal {
    id: ID!
    name: String
    age: Int
    address: String
  }
`;

const document = /* GraphQL */ `
  fragment person on Person {
    id
    name
    age
    address
  }

  fragment animal on Animal {
    id
    name
    age
    address
  }
`;

describe("configuration", () => {
  it("should be able to override the default faker method for String scalar", () => {
    const response = plugin(
      buildSchema(schema),
      [{ document: parse(document) }],
      {
        buildersOnly: true,
        scalars: { String: { _default: "myCustomStringFakerMethod()" } },
      },
    );
    expect(response).toMatchSnapshot();

    expect(response).toContain("name: myCustomStringFakerMethod()");
    expect(response).toContain("address: myCustomStringFakerMethod()");
  });

  it("should be able to override faker methods for specific field for String scalar", () => {
    const response = plugin(
      buildSchema(schema),
      [{ document: parse(document) }],
      {
        buildersOnly: true,
        scalars: { String: { name: "myCustomNameFakerMethod()" } },
      },
    );
    expect(response).toMatchSnapshot();

    expect(response).toContain("name: myCustomNameFakerMethod()");
    expect(response).toContain("address: faker.lorem.words()");
  });

  it("should be able to override faker methods for specific field on specific object for String scalar", () => {
    const response = plugin(
      buildSchema(schema),
      [{ document: parse(document) }],
      {
        buildersOnly: true,
        scalars: { String: { ["Person.name"]: "myCustomNameFakerMethod()" } },
      },
    );
    expect(response).toMatchSnapshot();

    expect(response).toContain("name: myCustomNameFakerMethod()");
    expect(response).toContain("name: faker.lorem.words()");
  });

  it("should be able to override faker methods according to specific configurations", () => {
    const response = plugin(
      buildSchema(schema),
      [{ document: parse(document) }],
      {
        buildersOnly: true,
        scalars: {
          String: {
            _default: "myCustomDefaultStringFakerMethod()",
            name: "myCustomNameFakerMethod()",
            ["Person.name"]: "myCustomPersonNameFakerMethod()",
          },
          ID: {
            _default: "myCustomDefaultIdStringFakerMethod()",
          },
        },
      },
    );
    expect(response).toMatchSnapshot();

    expect(response).toContain("name: myCustomPersonNameFakerMethod()");
    expect(response).toContain("name: myCustomNameFakerMethod()");
    expect(response).toContain("address: myCustomDefaultStringFakerMethod()");
    expect(response).toContain("id: myCustomDefaultIdStringFakerMethod()");
  });

  it("should be able to override faker seed", () => {
    const response = plugin(
      buildSchema(schema),
      [{ document: parse(document) }],
      {
        fakerjsSeed: 56765,
      },
    );
    expect(response).toMatchSnapshot();

    expect(response).toContain("faker.seed(56765)");
  });

  it("should be able to set location of the type import", () => {
    const response = plugin(
      buildSchema(schema),
      [{ document: parse(document) }],
      {
        typeImport: "@types",
      },
    );
    expect(response).toMatchSnapshot();

    expect(response).toContain("import * as types from '@types';");
  });

  it("should be able to skip fields", () => {
    const response = plugin(
      buildSchema(schema),
      [{ document: parse(document) }],
      {
        skipFields: ["PersonFragment.name", "AnimalFragment.name"],
      },
    );
    expect(response).toMatchSnapshot();

    expect(response).not.toContain("name: ");
  });

  it("should be able to skip fragments", () => {
    const response = plugin(
      buildSchema(schema),
      [{ document: parse(document) }],
      {
        skipFragments: ["person"],
      },
    );
    expect(response).toMatchSnapshot();

    expect(response).not.toContain("export const fakePerson");
  });
});
