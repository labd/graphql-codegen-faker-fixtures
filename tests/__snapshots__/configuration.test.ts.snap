// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`configuration > should be able to override faker methods according to specific configurations 1`] = `
"
    export const fakePerson = (override?:DeepPartial<types.PersonFragment>):types.PersonFragment => {
      const fixture:types.PersonFragment = {
        id: myCustomDefaultIdStringFakerMethod(),name: myCustomPersonNameFakerMethod(),age: faker.number.int(),address: myCustomDefaultStringFakerMethod()
      }

      return override
        ? merge(fixture, override)
        : fixture
    }
  

    export const fakeAnimal = (override?:DeepPartial<types.AnimalFragment>):types.AnimalFragment => {
      const fixture:types.AnimalFragment = {
        id: myCustomDefaultIdStringFakerMethod(),name: myCustomNameFakerMethod(),age: faker.number.int(),address: myCustomDefaultStringFakerMethod()
      }

      return override
        ? merge(fixture, override)
        : fixture
    }
  "
`;

exports[`configuration > should be able to override faker methods for specific field for String scalar 1`] = `
"
    export const fakePerson = (override?:DeepPartial<types.PersonFragment>):types.PersonFragment => {
      const fixture:types.PersonFragment = {
        id: faker.string.uuid(),name: myCustomNameFakerMethod(),age: faker.number.int(),address: faker.lorem.words()
      }

      return override
        ? merge(fixture, override)
        : fixture
    }
  

    export const fakeAnimal = (override?:DeepPartial<types.AnimalFragment>):types.AnimalFragment => {
      const fixture:types.AnimalFragment = {
        id: faker.string.uuid(),name: myCustomNameFakerMethod(),age: faker.number.int(),address: faker.lorem.words()
      }

      return override
        ? merge(fixture, override)
        : fixture
    }
  "
`;

exports[`configuration > should be able to override faker methods for specific field on specific object for String scalar 1`] = `
"
    export const fakePerson = (override?:DeepPartial<types.PersonFragment>):types.PersonFragment => {
      const fixture:types.PersonFragment = {
        id: faker.string.uuid(),name: myCustomNameFakerMethod(),age: faker.number.int(),address: faker.lorem.words()
      }

      return override
        ? merge(fixture, override)
        : fixture
    }
  

    export const fakeAnimal = (override?:DeepPartial<types.AnimalFragment>):types.AnimalFragment => {
      const fixture:types.AnimalFragment = {
        id: faker.string.uuid(),name: faker.lorem.words(),age: faker.number.int(),address: faker.lorem.words()
      }

      return override
        ? merge(fixture, override)
        : fixture
    }
  "
`;

exports[`configuration > should be able to override faker seed 1`] = `
"
    /* NOTE: This file is auto-generated. DO NOT EDIT. */
  

import * as types from 'CONFIGURE_TYPE_IMPORT';
import { deepmergeCustom } from 'deepmerge-ts'
import { faker } from '@faker-js/faker';


      export type DeepPartial<T> = T extends object
        ? { [P in keyof T]?: DeepPartial<T[P]> }
        : T

      export const repeat = <T>(numTimes: number, callback: () => T) =>
        Array.from(Array(numTimes).keys()).map(() => callback())

      const deepmerge = deepmergeCustom({ mergeArrays: false })
      export const merge = <T>(fixture: T, override: unknown) =>
        deepmerge(fixture, override) as T

      export const random = (min: number, max: number) => faker.number.int({ min, max })

      faker.seed(56765)
  


    export const fakePerson = (override?:DeepPartial<types.PersonFragment>):types.PersonFragment => {
      const fixture:types.PersonFragment = {
        id: faker.string.uuid(),name: faker.lorem.words(),age: faker.number.int(),address: faker.lorem.words()
      }

      return override
        ? merge(fixture, override)
        : fixture
    }
  

    export const fakeAnimal = (override?:DeepPartial<types.AnimalFragment>):types.AnimalFragment => {
      const fixture:types.AnimalFragment = {
        id: faker.string.uuid(),name: faker.lorem.words(),age: faker.number.int(),address: faker.lorem.words()
      }

      return override
        ? merge(fixture, override)
        : fixture
    }
  "
`;

exports[`configuration > should be able to override the default faker method for String scalar 1`] = `
"
    export const fakePerson = (override?:DeepPartial<types.PersonFragment>):types.PersonFragment => {
      const fixture:types.PersonFragment = {
        id: faker.string.uuid(),name: myCustomStringFakerMethod(),age: faker.number.int(),address: myCustomStringFakerMethod()
      }

      return override
        ? merge(fixture, override)
        : fixture
    }
  

    export const fakeAnimal = (override?:DeepPartial<types.AnimalFragment>):types.AnimalFragment => {
      const fixture:types.AnimalFragment = {
        id: faker.string.uuid(),name: myCustomStringFakerMethod(),age: faker.number.int(),address: myCustomStringFakerMethod()
      }

      return override
        ? merge(fixture, override)
        : fixture
    }
  "
`;

exports[`configuration > should be able to set location of the type import 1`] = `
"
    /* NOTE: This file is auto-generated. DO NOT EDIT. */
  

import * as types from '@types';
import { deepmergeCustom } from 'deepmerge-ts'
import { faker } from '@faker-js/faker';


      export type DeepPartial<T> = T extends object
        ? { [P in keyof T]?: DeepPartial<T[P]> }
        : T

      export const repeat = <T>(numTimes: number, callback: () => T) =>
        Array.from(Array(numTimes).keys()).map(() => callback())

      const deepmerge = deepmergeCustom({ mergeArrays: false })
      export const merge = <T>(fixture: T, override: unknown) =>
        deepmerge(fixture, override) as T

      export const random = (min: number, max: number) => faker.number.int({ min, max })

      faker.seed(12345654321)
  


    export const fakePerson = (override?:DeepPartial<types.PersonFragment>):types.PersonFragment => {
      const fixture:types.PersonFragment = {
        id: faker.string.uuid(),name: faker.lorem.words(),age: faker.number.int(),address: faker.lorem.words()
      }

      return override
        ? merge(fixture, override)
        : fixture
    }
  

    export const fakeAnimal = (override?:DeepPartial<types.AnimalFragment>):types.AnimalFragment => {
      const fixture:types.AnimalFragment = {
        id: faker.string.uuid(),name: faker.lorem.words(),age: faker.number.int(),address: faker.lorem.words()
      }

      return override
        ? merge(fixture, override)
        : fixture
    }
  "
`;

exports[`configuration > should be able to skip fields 1`] = `
"
    /* NOTE: This file is auto-generated. DO NOT EDIT. */
  

import * as types from 'CONFIGURE_TYPE_IMPORT';
import { deepmergeCustom } from 'deepmerge-ts'
import { faker } from '@faker-js/faker';


      export type DeepPartial<T> = T extends object
        ? { [P in keyof T]?: DeepPartial<T[P]> }
        : T

      export const repeat = <T>(numTimes: number, callback: () => T) =>
        Array.from(Array(numTimes).keys()).map(() => callback())

      const deepmerge = deepmergeCustom({ mergeArrays: false })
      export const merge = <T>(fixture: T, override: unknown) =>
        deepmerge(fixture, override) as T

      export const random = (min: number, max: number) => faker.number.int({ min, max })

      faker.seed(12345654321)
  


    export const fakePerson = (override?:DeepPartial<types.PersonFragment>):types.PersonFragment => {
      const fixture:types.PersonFragment = {
        id: faker.string.uuid(),age: faker.number.int(),address: faker.lorem.words()
      }

      return override
        ? merge(fixture, override)
        : fixture
    }
  

    export const fakeAnimal = (override?:DeepPartial<types.AnimalFragment>):types.AnimalFragment => {
      const fixture:types.AnimalFragment = {
        id: faker.string.uuid(),age: faker.number.int(),address: faker.lorem.words()
      }

      return override
        ? merge(fixture, override)
        : fixture
    }
  "
`;

exports[`configuration > should be able to skip fragments 1`] = `
"
    /* NOTE: This file is auto-generated. DO NOT EDIT. */
  

import * as types from 'CONFIGURE_TYPE_IMPORT';
import { deepmergeCustom } from 'deepmerge-ts'
import { faker } from '@faker-js/faker';


      export type DeepPartial<T> = T extends object
        ? { [P in keyof T]?: DeepPartial<T[P]> }
        : T

      export const repeat = <T>(numTimes: number, callback: () => T) =>
        Array.from(Array(numTimes).keys()).map(() => callback())

      const deepmerge = deepmergeCustom({ mergeArrays: false })
      export const merge = <T>(fixture: T, override: unknown) =>
        deepmerge(fixture, override) as T

      export const random = (min: number, max: number) => faker.number.int({ min, max })

      faker.seed(12345654321)
  


    export const fakeAnimal = (override?:DeepPartial<types.AnimalFragment>):types.AnimalFragment => {
      const fixture:types.AnimalFragment = {
        id: faker.string.uuid(),name: faker.lorem.words(),age: faker.number.int(),address: faker.lorem.words()
      }

      return override
        ? merge(fixture, override)
        : fixture
    }
  "
`;
